astropy/cosmology/tests/test_flrw.py

```diff
@@ -326,41 +316,79 @@ def test_m_nu(self, cosmo_cls, cosmo):
             assert u.allclose(cosmo.m_nu[:self._nmasslessnu], 0 * u.eV)
             assert u.allclose(cosmo.m_nu[self._nmasslessnu], cosmo._massivenu_mass)
 
-    def test_init_m_nu(self, cosmo_cls):
-        """Test initialization for values of ``m_nu``."""
-        ba = self.ba
+    def test_init_m_nu(self, cosmo_cls, ba):
+        """Test initialization for values of ``m_nu``.
 
-        # test that it works with units
+        Note this requires the class to have a property ``has_massive_nu``.
+        """
+        # Test that it works when m_nu has units.
         cosmo = cosmo_cls(*ba.args, **ba.kwargs)
-        assert np.all(cosmo.m_nu == ba.arguments["m_nu"])
+        assert np.all(cosmo.m_nu == ba.arguments["m_nu"])  # (& checks len, unit)
+        assert not cosmo.has_massive_nu
+        assert cosmo.m_nu.unit == u.eV  # explicitly check unit once.
 
-        # also without units
+        # And it works when m_nu doesn't have units.
         ba.arguments["m_nu"] = ba.arguments["m_nu"].value  # strip units
         cosmo = cosmo_cls(*ba.args, **ba.kwargs)
         assert np.all(cosmo.m_nu.value == ba.arguments["m_nu"])
+        assert not cosmo.has_massive_nu
 
-        # negative m_nu fails
+        # A negative m_nu raises an exception.
         tba = copy.copy(ba)
         tba.arguments["m_nu"] = u.Quantity([-0.3, 0.2, 0.1], u.eV)
-        with pytest.raises(ValueError):
+        with pytest.raises(ValueError, match="invalid"):
             cosmo_cls(*tba.args, **tba.kwargs)
 
-        # mismatch with Neff and Tcmb0
+    def test_init_m_nu_and_Neff(self, cosmo_cls, ba):
+        """Test initialization for values of ``m_nu`` and ``Neff``.
+
+        Note this test requires ``Neff`` as constructor input, and a property
+        ``has_massive_nu``.
+        """
+        # Mismatch with Neff = wrong number of neutrinos
         tba = copy.copy(ba)
-        tba.arguments["Tcmb0"] = 3
-        tba.arguments["Neff"] = 2
+        tba.arguments["Neff"] = 4.05
         tba.arguments["m_nu"] = u.Quantity([0.15, 0.2, 0.1], u.eV)
-        with pytest.raises(ValueError):
+        with pytest.raises(ValueError, match="unexpected number of neutrino"):
             cosmo_cls(*tba.args, **tba.kwargs)
 
-        # wrong number of neutrinos
+        # No neutrinos, but Neff
+        tba.arguments["m_nu"] = 0
+        cosmo = cosmo_cls(*tba.args, **tba.kwargs)
+        assert not cosmo.has_massive_nu
+        assert len(cosmo.m_nu) == 4
+        assert cosmo.m_nu.unit == u.eV
+        assert u.allclose(cosmo.m_nu, [0.0, 0.0, 0.0, 0.0] * u.eV)
```

Comment:
Should it not just be equal to zero?

Reply:
SGTM. I was more copying over the test for continuity. But improvement is better.